### ISSUE TITLE
Endless package upgrade.

### ISSUE THREAD

**Issue by fbourigault** [composer/composer#6537]
The [php-http/mock-client](https://packagist.org/packages/php-http/mock-client) package upgrade endlessly from `v1.0.1` to `v1.0.1`.

To reproduce the problem, you can clone the [php-http/httplug-bundle](https://github.com/php-http/HttplugBundle) project and run `composer update`. Then, run `composer update` again and you will get the following output:
```
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Updating php-http/mock-client (v1.0.1 => v1.0.1): Loading from cache
Generating autoload files
```

Running composer with PHP default configuration; cache disabled and ignoring cache (`COMPOSER_CACHE_DIR=/dev/null php -n ~/.local/bin/composer --ignore-platform-reqs update`) does not solve the problem.

I can't reproduce the issue by running `composer require php-http/mock-client` and `composer update` in an empty directory.

`composer diag` output:
```
Checking composer.json: OK
Checking platform settings: OK
Checking git settings: OK
Checking http connectivity to packagist: OK
Checking https connectivity to packagist: OK
Checking github.com oauth access: OK
Checking disk free space: OK
Checking pubkeys: 
Tags Public Key Fingerprint: 57815BA2 7E54DC31 7ECC7CC5 573090D0  87719BA6 8F3BB723 4E5D42D0 84A14642
Dev Public Key Fingerprint: 4AC45767 E5EC2265 2F0C1167 CBBB8A2B  0C708369 153E328C AD90147D AFE50952
OK
Checking composer version: OK
```

`php -v output`:
```
PHP 7.1.7 (cli) (built: Jul  5 2017 15:58:06) ( NTS )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.1.0, Copyright (c) 1998-2017 Zend Technologies
    with Zend OPcache v7.1.7, Copyright (c) 1999-2017, by Zend Technologies
    with Xdebug v2.5.4, Copyright (c) 2002-2017, by Derick Rethans
    with blackfire v1.16.2~linux-x64-non_zts71, https://blackfire.io, by Blackfireio Inc.
```

**Comment by fbourigault** [comment]
Dependency affected is not always the same but is still the same on consecutive calls.
```
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Updating php-http/react-adapter (v0.2.2 => v0.2.2): Loading from cache
Generating autoload files
```

**Comment by cabdesigns** [comment]
I seem to experiencing the same issue in Docker (PHP 7.0.21), either running composer install or self-update - just seems to hang at the end of the step.

**Comment by cabdesigns** [comment]
An update on this - seemed to be more of a Docker for Mac issue? Docker must have managed to get into a bad state, as I was unable to restart the daemon, even diagnostics looked fine. A restart of my machine seemed to resolve it all.

I can't however explain why it was hanging on composer commands in particular in the docker build, but I'm unblocked now.

**Comment by fbourigault** [comment]
I doesn't have any hang and I'm not running within docker. I use PHP from Arch Linux repositories.

**Comment by Seldaek** [comment]
Really weird.. I can reproduce it but no idea why it does this, as it updates between two packages which are entirely identical.

It might have something to do with those "provide" statements I suspect, but I can't really explain it sorry. It's probably going to be a while before we can look into it in more details too.

**Comment by barryvdh** [comment]
Probably the same issue as https://github.com/composer/composer/issues/6218 ?

**Comment by fbourigault** [comment]
Don't be sorry. It's something really minor and it does not bother me. I will try to look at it. It's a good opportunity to look at the composer internals.

**Comment by Seldaek** [comment]
Closing as seems to be fixed in 2.0, at least I can't reproduce, not sure if fixed or if the packages changed so much that it doesn't happen anymore.
